runtime/domain.c

```diff
@@ -297,6 +298,11 @@ Caml_inline void interrupt_domain(struct interruptor* s)
   atomic_store_release(s->interrupt_word, (uintnat)(-1));
 }
 
+Caml_inline dom_internal *internal_of_interruptor(struct interruptor *s)
+{
+  return (dom_internal *)((char *)s - offsetof(dom_internal, interruptor));
```

Comment:
I think this calls for a bigger reorganization to avoid such tricks. Have you considered other ways to make the parent domain accessible at domain creation?

Reply:
I am used to this `offsetof` technique from other work but I'm happy to do it another way. The easy alternative is to put the `caml_domain_state *` into the `domain_startup_params`.

Reply:
Or just the whole `dom_internal *` as `parent`, rather than two of its fields. Simplifies some code.

Reply:
I dug deeper as I was a bit puzzled as to why one maintains a pointer to another domain's state when access to this domain state is usually local-only for thread-safety reasons. This is where the `_Atomic stopped` in `memprof_config_s` comes into play as it is meant to be accessed in parallel.

My first worry is the following: while the domain state is allocated once and persists through domain reallocation (this is how `young_limit` can be changed from another domain without worries about memory safety), here the `memprof_config_s` (via `memprof_domain_s`) is allocated dynamically and deallocated upon domain shutdown. This makes me nervous about the thread-safety of accessing `stopped` from another domain. (To make things simpler, it seems that one could allocate the domain-local memprof data once and for all and recycle the allocated structure, like one does for the domain state.)

My second observation is that the memprof runtime has a concept of a parent-child relationship between domains whereas the domain runtime itself does not maintain such a relationship. This might explain why the code is a bit convoluted. I wonder how this parent-child relationship is important API-wise; it seems that only the relationship of having the same memprof "profile" is important. (Maybe I am reading too much into the variable names and this idea of parent-child does not affect the rest of the implementation.)

My last worry is that you use this parent domain state only (IIUC) at domain creation in order to duplicate the `stopped` mutable status and the config parameters. I am wondering how much things could be simpler if this atomic `stopped` switch was in fact on the OCaml heap and shared between domains (avoiding state replication). Pushing this idea further, could the `profile` consist of a heap-allocated OCaml value (comprising the `stopped` boolean and the rest of the data)? It looks to me as if this organisation could simplify things (and it will make sense to pass this profile value among the startup params). Do you have an insight to share that would clarify why things are done as they are now (notably regarding state replication of the `stopped` flag)?

Reply:
The parent/child relationship does not persist in the memprof data structures; it is only used to initialize memprof status for the new domain (by copying that from the parent domain).

Reply:
More detail: the memprof data structure for a child domain does not keep any reference to the parent domain's data structures. The configuration is copied, not pointed to. Memprof *will* keep track of the overall set of per-domain memprof structures, and will sometimes have to traverse this overall structure (for example, in order to set the per-domain `stopped` flag for one or more other domains). Such traversal and setting will be guarded by appropriate locks etc, for thread safety.

The reason for using a per-domain `stopped` flag, rather than a per-profile flag, is that it is checked frequently by its owning domain, only set at all when the profile is stopped, and only set by another domain when a profile is stopped which is shared between several domains. So in (what I think of as) the common case - in which domains do not share profiles - there should be very low cost to the atomicity (a domain's config can live in that domain's core's cache with no cache consistency traffic).

The idea of having a per-profile object on the Caml heap did occur to me. I didn't see a need for it, but we could migrate to such a system in the future, with only very localised code changes.